{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reference Transcriptome for Three _Abies_\n",
    "xanadu path: `/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes`\n",
    "gitlab: "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Tools:**\n",
    "* [Trinity (v 2.2.0)](https://github.com/trinityrnaseq/trinityrnaseq/wiki)\n",
    "* [GeneMarkS-T](http://topaz.gatech.edu/GeneMark/)\n",
    "* [VSEARCH (v 1.1.1)](https://github.com/torognes/vsearch)\n",
    "* [EnTAP](https://entap.readthedocs.io/en/latest/)\n",
    "* [OrthoFinder (v 1.1.8)](https://davidemms.github.io/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Transcriptomes:**\n",
    "* Fraser fir (_Abies fraseri_) \n",
    "* balsam fir (_Abies balsamea_)\n",
    "* Canaan fir (_Abies balsamea_ var. phanerolepis)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference Transcriptomes for each species\n",
    "1. A _de novo_ assembly for each library was constructed using [Trinity (v 2.2.0)](https://github.com/trinityrnaseq/trinityrnaseq/wiki). \n",
    "2. Frame selection of assembled contigs was executed using [GeneMarkS-T](http://topaz.gatech.edu/GeneMark/), a tool for ab initio identification of the protein coding regions in RNA transcripts. \n",
    "3. Individually, for each species, assembled transcripts from all libraries were concatenated into a single file. \n",
    "4. The contigs in the merged file were clustered at .90 identity using [VSEARCH (v 1.1.1)](https://github.com/torognes/vsearch)\n",
    "5. Functional annotation via [EnTAP](https://entap.readthedocs.io/en/latest/) (see methods below) identified sequences that aligned with confidence to fungal, bacterial, amoeba, and insect full length proteins in NCBIâ€™s RefSeq database. These contaminant transcripts were removed from the data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combined Reference Strategies\n",
    "Two strategies were employed to build reference transcriptomes which included transcripts from the three taxa. The first strategy clustered the three single species transciptomes together using VSEARCH. The second strategy carries out the same VSEARCH clustering, but does so after phylogenetic orthology inference with OrthoFinder. In the second strategy, each resulting orthogroup is input into VSEARCH separately, resulting in 64000 separate VSEARCH runs, one for each orthogroup. In both cases the centroids from all resulting VSEARCH clusters were taken to build reference transcriptomes. Experimental filters were applied to enable the resolution of patterns in gene expression that are common across the three taxa."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Orthofinder Strategy\n",
    "Using [OrthoFinder (v 1.1.8)](https://davidemms.github.io/), frame-selected transcripts from the three reference transcriptomes were clustered into orthologous gene families. click filenames below to view statistics from that run. Xanadu path: `/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting`\n",
    "\n",
    "<br>\n",
    "<details>    \n",
    "    <summary><b>Statistics_Overall.csv</b></summary>\n",
    "\n",
    "```\n",
    "Number of genes 439230\n",
    "Number of genes in orthogroups  256319\n",
    "Number of unassigned genes      182911\n",
    "Percentage of genes in orthogroups      58.4\n",
    "Percentage of unassigned genes  41.6\n",
    "Number of orthogroups   61400\n",
    "Number of species-specific orthogroups  122\n",
    "Number of genes in species-specific orthogroups 419\n",
    "Percentage of genes in species-specific orthogroups     0.1\n",
    "Mean orthogroup size    4.2\n",
    "Median orthogroup size  3.0\n",
    "G50 (assigned genes)    5\n",
    "G50 (all genes) 2\n",
    "O50 (assigned genes)    17725\n",
    "O50 (all genes) 43048\n",
    "Number of orthogroups with all species present  31076\n",
    "Number of single-copy orthogroups       4372\n",
    "\n",
    "Average number of genes per-species in orthogroup       Number of orthogroups   Percentage of orthogroups\n",
    "       Number of genes Percentage of genes\n",
    "<1      19585   31.9    39170   15.3\n",
    "'1      29273   47.7    114901  44.8\n",
    "'2      9007    14.7    60203   23.5\n",
    "'3      2252    3.7     21971   8.6\n",
    "'4      751     1.2     9595    3.7\n",
    "'5      268     0.4     4223    1.6\n",
    "'6      113     0.2     2126    0.8\n",
    "'7      64      0.1     1401    0.5\n",
    "'8      31      0.1     772     0.3\n",
    "'9      21      0.0     587     0.2\n",
    "'10     9       0.0     278     0.1\n",
    "11-15   21      0.0     768     0.3\n",
    "16-20   4       0.0     217     0.1\n",
    "21-50   1       0.0     107     0.0\n",
    "\n",
    "Number of species in orthogroup Number of orthogroups\n",
    "1       122\n",
    "2       30202\n",
    "3       31076\n",
    "```\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "    <summary><b>Statistics_PerSpecies.csv</b></summary>\n",
    "\n",
    "```\n",
    "        balsam.faa      canaan.faa       fraser.faa\n",
    "Number of genes 104193  125102  209935\n",
    "Number of genes in orthogroups  74793   80911   100615\n",
    "Number of unassigned genes      29400   44191   109320\n",
    "Percentage of genes in orthogroups      71.8    64.7    47.9\n",
    "Percentage of unassigned genes  28.2    35.3    52.1\n",
    "Number of orthogroups containing species        47783   52469   53502\n",
    "Percentage of orthogroups containing species    77.8    85.5    87.1\n",
    "Number of species-specific orthogroups  16      29      77\n",
    "Number of genes in species-specific orthogroups 49      88      282\n",
    "Percentage of genes in species-specific orthogroups     0.0     0.1     0.1\n",
    "\n",
    "\n",
    "Number of genes per-species in orthogroup       Number of orthogroups   Number of orthogroups   Number of orthogroups\n",
    "'0      13617   8931    7898\n",
    "'1      27833   36227   26587\n",
    "'2      16264   10074   17638\n",
    "'3      2124    3404    4516\n",
    "'4      891     1450    2366\n",
    "'5      306     596     1066\n",
    "'6      153     322     540\n",
    "'7      79      158     292\n",
    "'8      31      83      186\n",
    "'9      31      41      119\n",
    "'10     14      35      52\n",
    "11-15   42      59      103\n",
    "16-20   10      14      26\n",
    "21-50   5       6       11\n",
    "\n",
    "Number of genes per-species in orthogroup       Percentage of orthogroups       Percentage of orthogroups\n",
    "       Percentage of orthogroups\n",
    "'0      22.2    14.5    12.9\n",
    "'1      45.3    59.0    43.3\n",
    "'2      26.5    16.4    28.7\n",
    "'3      3.5     5.5     7.4\n",
    "'4      1.5     2.4     3.9\n",
    "'5      0.5     1.0     1.7\n",
    "'6      0.2     0.5     0.9\n",
    "'7      0.1     0.3     0.5\n",
    "'8      0.1     0.1     0.3\n",
    "'9      0.1     0.1     0.2\n",
    "'10     0.0     0.1     0.1\n",
    "11-15   0.1     0.1     0.2\n",
    "\n",
    "Number of genes per-species in orthogroup       Number of genes Number of genes Number of genes\n",
    "'0      0       0       0\n",
    "'1      27833   36227   26587\n",
    "'2      32528   20148   35276\n",
    "'3      6372    10212   13548\n",
    "'4      3564    5800    9464\n",
    "'5      1530    2980    5330\n",
    "'6      918     1932    3240\n",
    "'7      553     1106    2044\n",
    "'8      248     664     1488\n",
    "'9      279     369     1071\n",
    "'10     140     350     520\n",
    "11-15   526     732     1299\n",
    "16-20   173     247     458\n",
    "21-50   129     144     290\n",
    "\n",
    "Number of genes per-species in orthogroup       Percentage of genes     Percentage of genes     Percentage of genes\n",
    "'0      0.0     0.0     0.0\n",
    "'1      26.7    29.0    12.7\n",
    "'2      31.2    16.1    16.8\n",
    "'3      6.1     8.2     6.5\n",
    "'4      3.4     4.6     4.5\n",
    "'5      1.5     2.4     2.5\n",
    "'6      0.9     1.5     1.5\n",
    "'7      0.5     0.9     1.0\n",
    "'8      0.2     0.5     0.7\n",
    "'9      0.3     0.3     0.5\n",
    "'10     0.1     0.3     0.2\n",
    "11-15   0.5     0.6     0.6\n",
    "16-20   0.2     0.2     0.2\n",
    "21-50   0.1     0.1     0.1\n",
    "```\n",
    "\n",
    "</details>\n",
    "    \n",
    "To experiment with optimal clustering identity, orthogroups were then clustered at 90% and 85% using VSEARCH to split  divergent transcripts into independent clusters. For each clustering identity, centroids from every cluster were selected to consistute the combined reference transcriptome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### **Orthofinder Methods**\n",
    "```bash\n",
    "# Orthofinder.csv was split into 13 files of 5000 lines for parallel processing.\n",
    "split Orthogroups.csv -l 5000 Orthogroups\n",
    "```\n",
    "<details open>\n",
    "<summary>Submit 13 scripts to run of the split files.</summary>\n",
    "\n",
    "```bash\n",
    "for x in {a..m}; do if [[ -v PRE ]]; then sed -i \"s/Orthogroupsa$PRE/Orthogroupsa$x/g\" vsearch_single.sh && sbatch vsearch_single.sh && PRE=$x; else sbatch vsearch_single.sh && PRE=$x; fi; done\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "\n",
    "<details>\n",
    "<summary> <b>vsearch_single.sh</b> (click to open script) was run on 13 files.</summary>\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=Orthogroupsam\n",
    "#SBATCH -n 1\n",
    "#SBATCH -N 1\n",
    "#SBATCH -c 12\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --partition=himem\n",
    "#SBATCH --qos=himem\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --mail-user=alexander.trouern-trend@uconn.edu\n",
    "#SBATCH -o Orthogroupsam_%j.out\n",
    "#SBATCH -e Orthogroupsam_%j.err\n",
    "\n",
    "module load vsearch\n",
    "\n",
    "for TASK_ID in $(seq 1 $(wc -l Orthogroupsam | cut -f1 -d' '));\n",
    "do\n",
    "    echo \"host name : \" `hostname`\n",
    "    echo This is array task number $TASK_ID\n",
    "\n",
    "    BFPATH=\"/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting/balsam.fnn\"\n",
    "    CFPATH=\"/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting/canaan.fnn\"\n",
    "    FFPATH=\"/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting/fraser.fnn\"\n",
    "\n",
    "    # Gather data\n",
    "    OG=$(sed -n ${TASK_ID}p Orthogroupsam | cut -f 1 )\n",
    "    BF=$(sed -n ${TASK_ID}p Orthogroupsam | cut -f 2 | sed 's/, /TXXXT/g')\n",
    "    CF=$(sed -n ${TASK_ID}p Orthogroupsam | cut -f 3 | sed 's/, /TXXXT/g')\n",
    "    FF=$(sed -n ${TASK_ID}p Orthogroupsam | cut -f 4 | sed 's/, /TXXXT/g')\n",
    "\n",
    "    # Make a designated directory for orthogroup\n",
    "    cd output2\n",
    "\n",
    "    DIR=$\"id_${OG}\"\n",
    "\n",
    "    if [ ! -d \"$DIR\" ]\n",
    "    then mkdir \"$DIR\"\n",
    "    fi\n",
    "\n",
    "    cd \"$DIR\"\n",
    "\n",
    "    TRAN=($BFPATH $CFPATH $FFPATH)\n",
    "    LIST=(\"$BF\" \"$CF\" \"$FF\")\n",
    "\n",
    "    IDX=0 #initialize index\n",
    "    # run grabseqs to generate nucleotide multifasta for clustering\n",
    "    for i in \"${TRAN[@]}\";\n",
    "    do\n",
    "      python /labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting/grabseqs.py \\\n",
    "        -f $i \\\n",
    "        -l ${LIST[$IDX]} \\\n",
    "        -o ./${OG}.fnn\n",
    "      IDX=$((IDX+1));\n",
    "    done\n",
    "\n",
    "    # Cluster at 90% identity\n",
    "    vsearch --threads 12 --cluster_fast ${OG}.fnn --id 0.90 --centroids ${OG}_centroids.fnn --clusters ${OG}_;\n",
    "    cd ../../\n",
    "done\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "<summary> <b>grabseqs.py</b> (click to open script) is called by vsearch_single.sh.</summary>\n",
    "\n",
    "```python\n",
    "\n",
    "from Bio import SeqIO\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(\n",
    "        prog='grabseqs.py',\n",
    "        usage='''python grabseqs.py -f [Fasta file to subset] -l [list of headers representing desired subset] -o [output subset fasta]''', description='''This script produces a subet of a fasta file based on a text file containing sequence headers of interest.''',\n",
    "        epilog='''Requirements: Biopython library''')\n",
    "\n",
    "parser.add_argument(\"-f\", \"--fasta\", type=str, help=\"The fasta file to be subset\", required=True)\n",
    "parser.add_argument(\"-l\", \"--list\", type=str, help=\"A one-per-line textfile the headers to be taken from the fasta file.\", required=True)\n",
    "parser.add_argument(\"-o\", \"--out\", type=str, help=\"The resulting subsetted fasta file.\", required=True)\n",
    "\n",
    "args = parser.parse_args()\n",
    "fa = args.fasta\n",
    "li = args.list\n",
    "ou = args.out\n",
    "\n",
    "wanted = set()\n",
    "splt = li.split(\"TXXXT\")\n",
    "for line in splt:\n",
    "    line = line.strip()\n",
    "    if line != \"\":\n",
    "        wanted.add(line)\n",
    "\n",
    "fasta_sequences = SeqIO.parse(open(fa),'fasta')\n",
    "with open(ou, \"a\") as f:\n",
    "    for seq in fasta_sequences:\n",
    "        if seq.id in wanted:\n",
    "            SeqIO.write([seq], f, \"fasta\")\n",
    "            \n",
    "```\n",
    "\n",
    "</details>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### VSEARCH Strategy\n",
    "Reference transcriptomes for the three taxa were concatenated into a single file and experimentally clustered using VSEARCH at 85% and 90%. For each clustering identity, centroids from every cluster were selected to consistute the combined reference transcriptome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<details>\n",
    "    <summary> <b>Sanity Check</b> (click) - do # of seqs in constructed .fnn files match number in OG? <b> Yes, they do</b></summary>\n",
    "\n",
    "```\n",
    "atrend@xanadu-02 /labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting/output2\n",
    "$ grep -c \">\" */*.fnn | grep -v \"centroids\"\n",
    "id_OG0000000/OG0000000.fnn:107\n",
    "id_OG0000001/OG0000001.fnn:59\n",
    "id_OG0000002/OG0000002.fnn:54\n",
    "id_OG0000003/OG0000003.fnn:53\n",
    "id_OG0000004/OG0000004.fnn:51\n",
    "id_OG0000005/OG0000005.fnn:45\n",
    "id_OG0000006/OG0000006.fnn:41\n",
    "id_OG0000007/OG0000007.fnn:41\n",
    "id_OG0000008/OG0000008.fnn:39\n",
    "\n",
    "atrend@xanadu-02 /labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting\n",
    "$ grep -o -n 'CF\\|Fraser\\|Balsam' Orthogroups.csv |  cut -d : -f 1 | uniq -c | head -n 10\n",
    "      1 1\n",
    "    107 2\n",
    "     59 3\n",
    "     54 4\n",
    "     53 5\n",
    "     51 6\n",
    "     45 7\n",
    "     41 8\n",
    "     41 9\n",
    "     39 10\n",
    "\n",
    "```\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strategy Assessment\n",
    "Clustering statistics were calculated using the custom python script, [clusterStats.py](https://github.com/alextrouerntrend/fir-transcriptomics/blob/master/scripts/clustStats.py), which searches the directories created by VSEARCH clustering by [vsearch_single.sh](https://github.com/alextrouerntrend/fir-transcriptomics/blob/master/scripts/vsearch_single.sh) and creates a stats table like [o85.tsv](https://github.com/alextrouerntrend/fir-transcriptomics/blob/master/data/transcriptome_stats/o85.tsv).\n",
    "<details open>\n",
    "    <summary><b>clusterStats.py</b></summary>\n",
    "\n",
    "```Python\n",
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Thu Jul 16 14:08:13 2020\n",
    "\n",
    "@author: Alex Trouern-Trend\n",
    "\"\"\"\n",
    "from Bio import SeqIO\n",
    "from statistics import mean, variance\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "df = pd.DataFrame(columns=['ClustID', 'seqCount', 'avgLen', 'lenVariance', 'species', 'fraser', 'canaan', 'balsam'])\n",
    "\n",
    "with open(\"fir90_centroids.fnn\") as f:\n",
    "     centroids = [ rec.id for rec in SeqIO.parse(f, \"fasta\")]\n",
    "\n",
    "for filename in sorted(os.listdir(\"fir90\")):\n",
    "    filename = \"fir90/\" + filename\n",
    "    cluster_dirty = sorted(os.listdir(filename))\n",
    "    print(cluster_dirty)\n",
    "    for item in cluster_dirty:\n",
    "        if item.endswith(\".fnn\"):\n",
    "            cluster_dirty.remove(item)\n",
    "    for cluster in cluster_dirty:\n",
    "        cluster = filename + \"/\" + cluster\n",
    "        cluname = cluster.split(\"/\")[-1]\n",
    "        print(cluster)\n",
    "        print(cluname)\n",
    "        sizes = [len(rec) for rec in SeqIO.parse(cluster, \"fasta\")]\n",
    "        headers = [rec.id for rec in SeqIO.parse(cluster, \"fasta\")]\n",
    "        total = len(sizes)\n",
    "        lenmean = mean(sizes)\n",
    "        if len(sizes) > 1:\n",
    "            vari = variance(sizes)\n",
    "        else:\n",
    "            vari = \"NULL\"\n",
    "        ff = 0\n",
    "        cf = 0\n",
    "        bf = 0\n",
    "        spec = \"\"\n",
    "        for i in headers:\n",
    "            if i in centroids:\n",
    "                centr = i\n",
    "            if i.startswith(\"F\"):\n",
    "                ff += 1\n",
    "                if ff == 1:\n",
    "                    spec += \"F\"\n",
    "            elif i.startswith(\"C\"):\n",
    "                cf += 1\n",
    "                if cf == 1:\n",
    "                    spec += \"C\"\n",
    "            elif i.startswith(\"B\"):\n",
    "                bf += 1\n",
    "                if bf == 1:\n",
    "                    spec += \"B\"\n",
    "        spec = ''.join(sorted(spec))\n",
    "        perF = ff/total\n",
    "        perC = cf/total\n",
    "        perB = bf/total\n",
    "        df = df.append({'ClustID' : cluname , 'seqCount' : total , 'centroid' : centr, 'avgLen' : lenmean , 'lenVariance' : vari , 'species' : spec , 'fraser' : perF , 'canaan' : perC , 'balsam' : perB}, ignore_index=True)\n",
    "df.to_csv(\"vsearchStats90.tsv\", sep='\\t', encoding='utf-8')    \n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "The tables were processed in RStudio using transcStats.R, which applied two cluster membership filters to determine cluster retention.\n",
    "1. Cluster must include >1 species.\n",
    "2. Cluster must include fraser fir transcript.\n",
    "\n",
    "<details open>\n",
    "    <summary><b>transcStats.R</b></summary>\n",
    "\n",
    "```R\n",
    "# Plant Computational Genommics lab @ UConn\n",
    "# Author: Alex Trouern-Trend\n",
    "# 2020-07-20\n",
    "\n",
    "#### BACKGROUND\n",
    "#####\n",
    "# TRANSCRIPTOME STATISTICS\n",
    "# Two methods of transcriptome construction by two clustering minimum identity.\n",
    "# \n",
    "# Clustering methods\n",
    "# 1. VSEARCH (single species) --> VSEARCH (multispecies)\n",
    "#    xanadu path: `/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch`\n",
    "# 2. VSEARCH (single species) --> Orthofinder (multispecies) --> VSEARCH (orthogroups)\n",
    "#    xanadu path: `/labs/Wegrzyn/Transcriptomics/xmas_trees/NAR/poola/transcriptomes/vsearch_splitting'\n",
    "#\n",
    "# Clustering identities (VSEARCH (multispecies) and VSEARCH (orthogroups)), VSEARCH (single species) non-variable at 90%\n",
    "# 1. 90%\n",
    "#    xanadu path: $CLUST_PATH`/fir90`\n",
    "# 2. 85%\n",
    "#     xanadu path: $CLUST_PATH`/fir85`\n",
    "#\n",
    "# Input .tsv files generated from clustStats.py (https://github.com/alextrouerntrend/bioscripts/blob/master/utils/clustStats.py)\n",
    "#     Data available at: \n",
    "#####\n",
    "\n",
    "library(dplyr)\n",
    "library(readr)\n",
    "library(knitr)\n",
    "library(stringr)\n",
    "\n",
    "# load in the datasets from github (will take a moment)\n",
    "link_v85 <- 'https://raw.github.com/alextrouerntrend/fir-transcriptomics/master/data/transcriptome_stats/v85.tsv'\n",
    "link_v90 <- 'https://raw.github.com/alextrouerntrend/fir-transcriptomics/master/data/transcriptome_stats/v90.tsv'\n",
    "link_o85 <- 'https://raw.github.com/alextrouerntrend/fir-transcriptomics/master/data/transcriptome_stats/o85.tsv'\n",
    "link_o90 <- 'https://raw.github.com/alextrouerntrend/fir-transcriptomics/master/data/transcriptome_stats/o90.tsv'\n",
    "\n",
    "v85 <- read.csv(link_v85, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "v90 <- read.csv(link_v90, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "o85 <- read.csv(link_o85, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "o90 <- read.csv(link_o90, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "\n",
    "# trim path from clustID\n",
    "pathtrim85 <- substring(v85$ClustID, first = 7)\n",
    "pathtrim90 <- substring(v90$ClustID, first = 7)\n",
    "v85$ClustID <- pathtrim85\n",
    "v90$ClustID <- pathtrim90\n",
    "\n",
    "# Filtering requirements\n",
    "t <- c(\"v85\", \"v90\", \"o85\", \"o90\")\n",
    "tfilt <- paste(t, \"_filt\", sep=\"\")\n",
    "\n",
    "for (i in t) {\n",
    "  p <- get(i)\n",
    "  p <- p[-1]\n",
    "  assign(i, p)\n",
    "}\n",
    "\n",
    "idx = 1\n",
    "for (i in t) {\n",
    "  assign(tfilt[idx], get(i) %>%\n",
    "  filter(str_detect(species, 'F')) %>%\n",
    "  filter(nchar(species) > 1))\n",
    "  idx <- idx + 1\n",
    "}\n",
    "# Statistical summary\n",
    "header <- c(\"transcriptome size\", \"mean cluster size\", \"mean seq length (full clusters)\", \"mean length variance\", \"mean % fraser\", \"mean % canaan\", \"mean % balsam\")\n",
    "\n",
    "tns <- c(t, tfilt)\n",
    "tnsStats <- paste(tns, \"_stats\", sep=\"\")\n",
    "\n",
    "\n",
    "idx = 1\n",
    "for (i in tns) {\n",
    "  p <- get(i)\n",
    "  assign(tnsStats[idx], c(length(p$ClustID), mean(p$seqCount), mean(p$avgLen), mean(as.numeric(p$lenVariance), na.rm = TRUE), mean(p$fraser), mean(p$canaan), mean(p$balsam)))\n",
    "  idx <- idx + 1\n",
    "}\n",
    "\n",
    "# Build summary table\n",
    "vXstat <- as.data.frame(rbind(v85_stats, v90_stats, o85_stats, o90_stats, v85_filt_stats, v90_filt_stats, o85_filt_stats, o90_filt_stats))\n",
    "colnames(vXstat) <- header\n",
    "rn <- c(\"vsearch85\", \"vsearch90\", \"orthofinder85\", \"orthofinder90\")\n",
    "rn2 <- c(rn, paste(rn, \"_filt\", sep=\"\"))\n",
    "rownames(vXstat) <- rn2\n",
    "\n",
    "# export summary table\n",
    "write.csv(vXstat, file = \"~/Documents/scientist/projects/xmas_NAR/fir_transcriptomics/data/transcriptome_stats/tn_summary.csv\", quote = F, row.names = T)\n",
    "```\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
